services/ingestion/subscriber_test.go

```diff
@@ -59,3 +64,412 @@ func Test_Subscribing(t *testing.T) {
 	// this makes sure we indexed all the events
 	require.Equal(t, uint64(endHeight), prevHeight)
 }
+
+func Test_SubscribingWithRetryOnError(t *testing.T) {
+	endHeight := uint64(10)
+	sporkClients := []access.Client{}
+	currentClient := testutils.SetupClientForRange(1, endHeight)
+
+	cadenceHeight := uint64(5)
+	txCount := 10
+	hashes := make([]gethCommon.Hash, txCount)
+	flowEvents := make([]flow.Event, 0)
+
+	// generate txs
+	for i := 0; i < txCount; i++ {
+		cdcEvent, txEvent, tx, _, err := newTransaction(cadenceHeight)
+		require.NoError(t, err)
+		hashes[i] = tx.Hash()
+		flowEvent := flow.Event{
+			Type:  string(txEvent.Etype),
+			Value: cdcEvent,
+		}
+		flowEvents = append(flowEvents, flowEvent)
+	}
+
+	evmTxEvents := flow.BlockEvents{
+		BlockID:        flow.Identifier{0x1},
+		Height:         cadenceHeight,
+		BlockTimestamp: time.Now(),
+		Events:         flowEvents,
+	}
+
+	// generate single block
+	cdcEvent, evmBlock, blockEvent, err := newBlock(cadenceHeight, hashes)
+	require.NoError(t, err)
+	flowEvent := flow.Event{
+		Type:  string(blockEvent.Etype),
+		Value: cdcEvent,
+	}
+	evmBlockEvents := flow.BlockEvents{
+		BlockID:        flow.Identifier{0x1},
+		Height:         cadenceHeight,
+		BlockTimestamp: time.Now(),
+		Events:         []flow.Event{flowEvent},
+	}
+
+	currentClient.On(
+		"GetEventsForHeightRange",
```

Comment:
could we extract all this setup section into a helper so all tests don't have to implement it

Reply:
Very good point :+1: It was a lot of duplicated code.
Updated in https://github.com/onflow/flow-evm-gateway/pull/529/commits/3c9ead3de7e2a226d0244c2860a2aafa8342e7a5